[Feat] GET 요청에 대한 accessToken , refreshToken 재발급 로직 생성

yonghyeun opened the pull request:
# 관련 이슈 번호

close #248 

# 설명

우선 백엔드 분과 협의 하에 각 리스폰스 별 `message` 부분을 통일 시켜, `message` 를 이용 한 분기 처리가 가능하게 되었음을 알려 드립니다. 

> `query` 에 대한 `onError` 핸들링은 `queryCache` 를 이용,  `mutation` 에 대한 에러 핸들링은 `defaultOptions` 에서 정의 합니다.
> 이번 작업에선 `query` 에 대한 핸들링만 처리하도록 합니다.
>
> `query` 는 `queryCache` 를 이용하는 이유는 #248 에 적은 작업 전에 정리 한 문서에 남겨둔 출처를 통해 확인 할 수 있습니다. 버전 5이상부턴 `query` 에서 `onError` 메소드를 사용하지 못하기 때문입니다. 

> 이전 작업에선 몰랐는데 이전 작업에서 `useGetProfile` 에서 토큰을 헤더에 집어넣지 않고 있었더군요 🤯 
>
> 반성합니다 .. 

## `queryClient` 생성 방식 변경 

```tsx

export const useCreateQueryClient = () => {
  const setToken = useAuthStore((state) => state.setToken);
  const resetAuthStore = useAuthStore((state) => state.reset);

  const queryClient = useRef(
    new QueryClient({
      defaultOptions: {
        queries: {
          /**
           * 기본 retry 는 false 로 해둡니다.
           * Access, RefreshToken 에러 발생 시, 불필요한 retry 를 막기 위함입니다.
           * 만약 retry를 필요로 하는 경우에선 개별적인 query 문에서 retry 를 정의 해줍니다.
           */
          retry: false,
        },
        mutations: {
          /**
           * 기본 retry 는 false 로 해둡니다.
           * Access, RefreshToken 에러 발생 시, 불필요한 retry 를 막기 위함입니다.
           * 만약 retry를 필요로 하는 경우에선 개별적인 query 문에서 retry 를 정의 해줍니다.
           */
          retry: false,
        },
      },

      queryCache: new QueryCache({
        onError: async (error, query) => {
          switch (error.message) {
            case ERROR_MESSAGE.ACCESS_TOKEN_INVALIDATED: {
              await getNewAccessToken({
                query,
                setterMethods: { setToken, resetAuthStore },
                queryClient,
              });
              break;
            }
            default:
              throw error;
          }
        },
      }),
    }),
  ).current;

  return queryClient;
};
```
```tsx
import { useCreateQueryClient } from "./useCreateQueryClient";

export const ReactQueryProvider = ({
  children,
}: {
  children: React.ReactNode;
}) => {
  const queryClient = useCreateQueryClient();

  return (
    <QueryClientProvider client={queryClient}>
      {children}
      <ReactQueryDevtools initialIsOpen={false} />
    </QueryClientProvider>
  );
};
```
쿼리클라이언트를 이제 전역이 아닌 `useCreateQueryClient` 훅을 이용해 생성하고 하위로 내려주도록 변경해주었습니다.

커스텀 훅을 통해 쿼리 클라이언트를 생성해주는 이유는 `queryClient` 의 `QueryCache` 에서 `use .. ` 로 시작하는 로직들을 소모하기 위함입니다. 

이번에 변경 사항을 보면 `queryCache` 에서 `onError`에 대해 조건문을 처리하고 내부에서 정의되는 에러 핸들링 메소드들에 `query, setterMethods , queryClient` 를 인수로 보내주는 모습을 볼 수 있습니다.

## 에러 핸들러 부착

```tsx
// app/QueryClient/errorHandlers.ts
import type { Query, QueryClient, QueryKey } from "@tanstack/react-query";
import { AuthStore } from "@/shared/store";
import { END_POINT, ERROR_MESSAGE } from "./constants";

export const getNewAccessToken = async ({
  query,
  setterMethods,
  queryClient,
}: {
  query: Query<unknown, unknown, unknown, QueryKey>;
  setterMethods: {
    setToken: AuthStore["setToken"];
    resetAuthStore: AuthStore["reset"];
  };
  queryClient: QueryClient;
}) => {
  const { queryKey } = query;
  const { setToken, resetAuthStore } = setterMethods;

  // 해당 try-catch 문은 access token 을 refresh token 을 이용해 재발급 받는 로직입니다.
  try {
    const response = await fetch(END_POINT.REFRESH_ACCESS_TOKEN, {
      /**
       * 개발 모드일 경우에만 쿠키를 같은 origin 으로만 전송합니다.
       * 그 이유는 개발 모드에서는 서버와 클라이언트가 다른 포트에서 동작하기 때문입니다.
       * 이로 인해 쿠키가 전송되지 않아 인증이 되지 않는 문제가 발생합니다.
       * 쿠키를 다른 origin 에게도 전송하게 되면 쿠키가 탈취되어 보안에 취약해질 수 있습니다.
       */
      credentials:
        process.env.NODE_ENV === "development" ? "include" : "same-origin",
    });
    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.message);
    }

    // 새로운 access token 을 AuthStore 에 저장합니다.
    // 토큰을 쿼리 키로 갖는 쿼리들은 자연스럽게 새로운 토큰을 사용하게 됩니다.
    setToken(data.content.authorization);
  } catch (error) {
    if (
      error instanceof Error &&
      error.message === ERROR_MESSAGE.REFRESH_TOKEN_INVALIDATED
    ) {
      /**
       * refresh 토큰이 유효하지 않다면 다음과 같은 일들을 처리 합니다.
       * 1. AuthStore 의 token, role, nickname 을 초기화 합니다.
       * 2. queryClient 의 queryKey 를 통해 해당 쿼리를 취소합니다. 리소스를 절약하고 잘못된 데이터가 저장되는 행위를 방지하기 위함입니다.
       * 3. 캐싱 된 쿼리를 모두 초기화 합니다.
       */
      resetAuthStore();
      queryClient.cancelQueries({ queryKey: queryKey });
      queryClient.clear();

      // 스토리북 환경의 경우엔 이하 코드를 실행 시키지 않습니다.
      // 스토리북은 react-router-dom을 사용하지 않기 때문입니다.
      if (
        (
          window as Window &
            typeof globalThis & { __STORYBOOK_ADDONS_CHANNEL__?: unknown }
        ).__STORYBOOK_ADDONS_CHANNEL__
      ) {
        return;
      }

      // 현재 ReactQueryProvider 는 ReactRouterDom 내부에 존재하지 않기에 window 를 이용해 직접 라우팅 시킵니다.
      // TODO : 추후 ReactQueryProvider 를 ReactRouterDom 내부로 이동시키고, useNavigate 를 이용해 라우팅 시키도록 수정합니다.
      // window.location.href = ROUTER_PATH.LOGIN;
      return;
    }

    throw error;
  }
};
```

`토큰 검증이 실패했습니다.` 라는 에러 메시지 (401 코드와 액세스 토큰 검증 실패) 가 발생하면 호출 될 에러 핸들러 메소드 입니다.

에러 핸들러 메소드 안에선 `try catch` 문과 함께 `refresh token` 을 재발급 합니다. 

이 때 `refresh token` 재발급이 성공하면 토큰을 새로운 값으로 업데이트 하고, 리프레쉬 토큰이 `stale` 하여 재발급이 실패 하면 캐싱 된 모든 쿼리를 지우고 로그아웃 처리 합니다. 

> 현재 `QueryClient` 는 react-router-dom 프로바이더 외부에 존재하여 `useNavigate` 를 사용하지 못합니다.
> 만약 `window.location.href` 를 하게 되면 `blink` 현상과 함께 서버에게 `/` 경로에 대한 요청을 다시 보내게 됩니다.
> 해당 풀리퀘가 머지 된 이후에는 `react-router-dom` 의 전체 레이아웃을 한 번 정리 한 후에 리액트 쿼리 프로바이더를 `react-router-dom` 안에서 정의해줘야 합니다. 

## 쿼리 키에 토큰도 같이 넣기 

이제 토큰이 변경 되면 새롭게 요청이 일어나야 하는 쿼리에 대해 쿼리 키에 토큰을 함께 넣어 줍니다.

```tsx
export const useGetProfile = ({ nickname, token }: ProfileRequest) => {
  return useQuery({
    queryKey: ["profile", nickname, token],
    queryFn: () => getProfile({ nickname, token }),

    enabled: !!nickname,
    staleTime: 1000 * 60 * 10, // 10분간 데이터는 캐시된 상태로 사용
    select: (data) => data?.content,
  });
};
```

이제 에러 핸들러에 의해 새로운 액세스 토큰이 발급 되면 `useGetProfile` 은 새로운 토큰과 함께 데이터를 패칭해오게 됩니다. 

## DevTools & msw 추가 

### DevTools

![image](https://github.com/user-attachments/assets/1ec10c82-df88-4bf9-8099-0add1df6d58d)

`Vite` 에서 테스트 하기 위해 두 가지 `DevTools` 와 `msw` 핸들러를 추가 했습니다.
 
`staleAT&freshRT` 는 리프레쉬 토큰은 유효하지만 액세스 토큰은 stale 한 상태

`staleAT&staleAT` 은 두 토큰 모두 `stale` 한 상태 입니다.

### msw 

```tsx
export const getProfileHandlers = [
  http.get(
    `${import.meta.env.VITE_API_BASE_URL}/profile`,
    async ({ request }) => {
      const requestUrl = new URL(request.url);
      const nickname = requestUrl.searchParams.get("nickname");
      // 2024/10/05 AccessToken 검증 로직을 추가 합니다.
      const token = request.headers.get("Authorization");
      if (token === "staleAccessToken") {
        return HttpResponse.json(
          {
            code: 401,
            message: ERROR_MESSAGE.ACCESS_TOKEN_INVALIDATED,
          },
          {
            status: 401,
          },
        );
      }
      if (token === "freshAccessToken" && nickname === "뽀송송") {
        return HttpResponse.json(User["ROLE_USER"]);
      }
```

특정 토큰들 (`stale , refreshAccessToken`)에 대해서 401 혹은 200 OK 를 보내는 msw 핸들러를 추가 합니다. 

## 사용 되는 모습

### Stale AT & fresh RT

![image](https://github.com/user-attachments/assets/d93ee6ce-29bc-4d60-9522-68347e1d0789)

`RefreshToken` 이 유효하면 `auth` 에서 액세스 토큰을 새롭게 업데이트 하고 새롭게 업데이트 한 액세스 토큰을 이용해 `useGetProfile` 이 호출됩니다.

### Stale AT & Stale RT 

![image](https://github.com/user-attachments/assets/cec44758-2b26-47ae-83ff-649b482d781f)

`auth` 에서도 401 에러가 발생하면 로그아웃 처리 하고 쿼리 초기화 합니다. 

> 다시 말 하지만 리팩토링 이후  `useNavigate`  이후 로그인 페이지로 리다이렉션 시켜야 합니다.












# 첨부 파일 (Optional)

# 레퍼런스 (Optional)

yonghyeun committed:
fix[#247] : useGetProfile 에서 token 도 인수로 받도록 수정

yonghyeun committed:
feat[#247] : MyPage access token 유효기간에 따른 실패 테스트 코드 작성

yonghyeun committed:
feat[#247] : getNewAccessToken 핸들러 생성

yonghyeun committed:
feat[#247] : useCreateQueryClient 를 통해 queryClient 생성

yonghyeun committed:
feat[#247] : 리액트 쿼리 에러 핸들리에 사용 할 상수 생성

yonghyeun committed:
feat[#247] : getProfileHandler 에서 액세스 토큰 검증 로직 추가

yonghyeun committed:
feat[#247] : 에러 핸들러 스토리북이 아닌 handler 로 이동

- 스토리북에선 cookie 설정이 원활하게 되지 않더라고요

yonghyeun committed:
refactor[#247] : 중복된 authorization 헤더 제거

yonghyeun committed:
feat[#247] : stale Access Token DevTools 추가

yonghyeun committed:
feat[#247] : msw handler 에서 access token 을 이용한 getMyProfile 핸들러 추가

yonghyeun committed:
feat[#247] : 토큰 값에 따라 재요청 받아야 하는 경우엔 쿼리키에 token 추가

yonghyeun committed:
feat[#247] : 에러 핸들러가 부착된 queryClient 를 반환하는 useCreateQueryClient 생성

yonghyeun committed:
refactor[#247] : AccessToken 톤에 맞춰 Access 란 단어 추가

yonghyeun committed:
refactor[#247] : 지우지 못했던 console.table 제거

yonghyeun committed:
feat[#247] : useGetProfile argument 변경에 따른 코드 변화 적용

yonghyeun committed:
Merge branch 'dev' into feat/#248-token-api

aken-you commented on src/entities/profile/api/dashboard.ts:
```diff
@@ -46,10 +47,23 @@ interface ProfileResponse {
-): Promise<ProfileResponse> => {
-  const response = await fetch(PROFILE_END_POINT.PROFILE(nickname));
+interface ProfileRequest extends Pick<AuthStore, "token"> {
+  nickname: string | null;
```
AuthStore에도 nickname이 있는데 따로 타입을 지정해준 이유가 있을까요 ?

yonghyeun replied:
`useGetProfile` 의 경우엔 마이 프로필  말고 타인의 프로필을 조회 할 때도 사용 될 예정 입니다. 

`useAuthStore` 는 실제 사용자의 개인 정보를 가리키는 역할이라 생각하여서 `AuthStore` 를 사용하지 않고  `string` 을 써줬습니다.

제 생각에 `useGetProfile` 의 인수로 들어가는 `token` 은 `AuthStore` 의 토큰이지만 , 닉네임은 정말 그냥 검색어라고 생각했어요

aken-you commented on src/app/ReactQueryProvider/errorHandlers.ts:
```diff
@@ -0,0 +1,75 @@
+       * 쿠키를 다른 origin 에게도 전송하게 되면 쿠키가 탈취되어 보안에 취약해질 수 있습니다.
+       */
+      credentials:
+        process.env.NODE_ENV === "development" ? "include" : "same-origin",
```
이전 풀리퀘에서 개발 모드에서 서버랑 클라이언트가 동일한 도메인을 사용하는 것으로 가정하고,
배포할 때 include로 할지 결정하자고 이야기 했던 것 같아요!

그럼 이 부분도 우선 same-origin으로 설정하는 걸로 하는걸까요?

yonghyeun replied:
넵 이 부분은 날려버린 커밋에서 수정했었던건데 반영이 안됐네요 반영하겠습니다

aken-you commented on .storybook/preview.tsx:
```diff
@@ -2,14 +2,17 @@ import { initialize, mswLoader } from "msw-storybook-addon";
 
-const queryClient = new QueryClient();
+const WithProviders = (Story, context) => {
+  // const navigate = useNavigate();
```
navigate는 나중에 사용되는 걸까요?

yonghyeun replied:
네비게이트는 왜 여기있지 .. 아마 코드를 복사해서 넣던 중 의도치 않게 같이 들어간 거 같습니다 🫠 제거할게요

aken-you commented on src/entities/profile/api/dashboard.ts:
```diff
@@ -58,13 +72,13 @@ export const getProfile = async (
+    queryKey: ["profile", nickname, token],
+    queryFn: () => getProfile({ nickname, token }),
 
     enabled: !!nickname,
```
postman 보니까 token이 있는 경우에만 요청을 보내는 거 같은데 맞을까요?!
그렇다면 enabled에 token이 없을 경우도 고려해봐야 할 것 같아요!

yonghyeun replied:
`useGetProfile` 의 경우에는 내 토큰을 이용해 내 정보 외에 다른 이의 정보도 함께 요청 할 때 사용하는 쿼리 입니다.

토큰이 존재하지 않아도 다른 이의 정보를 요청 할 수 있으니 , 토큰은 안넣어도 되지 않을까요 ? 🫠

aken-you replied:
지금 저희 기획에서 다른 사람 프로필을 접근하려면 적어도 guest 회원 이상이어야 하더라구요!
토큰이 존재하지 않으면 profile 정보를 접근할 수 없는 걸로 알고 있습니다.

<img width="458" alt="스크린샷 2024-10-07 오후 2 06 35" src="https://github.com/user-attachments/assets/02d2cf29-b7ba-43fd-8a5f-cbba923d7cfd">

yonghyeun replied:
저는 토큰이 존재하지 않으면 남의 프로필을 진입 할 수 없다는게 저는 논리적으로 이해가 가지 않습니다 .. 

이 부분은 논의 해보고 다른 브랜치에서 getProfile 만을 다시 수정하는 방향으로 하시죠

aken-you commented on src/entities/profile/api/dashboard.ts:
```diff
@@ -46,10 +47,23 @@ interface ProfileResponse {
+  nickname,
+  token,
+}: ProfileRequest): Promise<ProfileResponse | null> => {
+  if (!nickname) {
+    return null;
+  }
```
nickname 값이 null이면 getProfile을 호출하지 않도록 할 수 있었을 것 같은데,
여기서 처리하신 이유가 있을까요 ?

yonghyeun replied:
이 부분은 쿼리를 호출하는 장소에 대한 고민을 해결한 후 다른 풀리퀘에서 작업해보시죠 

getProfile API 에 대한 부분도 백엔드와 이야기를 해봐야겠습니다

aken-you commented on src/app/ReactQueryProvider/useCreateQueryClient.ts:
```diff
@@ -0,0 +1,52 @@
+        },
+      },
+
+      queryCache: new QueryCache({
```
QueryCache 처음 봤는데 이번 기회에 배워갑니다 👍

aken-you commented on src/app/ReactQueryProvider/errorHandlers.ts:
```diff
@@ -0,0 +1,75 @@
+       */
+      resetAuthStore();
+      queryClient.cancelQueries({ queryKey: queryKey });
+      queryClient.clear();
```
cancelQueries를 쓰지 않더라도 clear로 인해 모든 쿼리가 모두 취소되고 캐싱된 데이터도 삭제되는데, cancelQueries를 사용한 이유가 있을까요 ?

yonghyeun replied:
우선 `DevTools` 에서 Auth 내용이 변경 될 때 마다 쿼리 데이터를 `setQueryData , invalidated ... ` 하는 과정은 모두 제거 되었습니다.

이제는 `msw` 핸들러를 통해 실제로 데이터를 받아오니까요 !

그나저나 공부 차 실험해봤습니다.

```tsx
  const handleClick = () => {
    queryClient.setQueryData(["experiment"], { data: "staleData" });
    queryClient.fetchQuery({
      queryKey: ["experiment"],
      queryFn: () =>
        new Promise((res) =>
          setTimeout(() => res({ data: "still stale data" }), 1000),
        ),
    });
    queryClient.clear();
  };
```

다음과 같이 실험했을 때 클리어를 하게 되면 클리어 이전에 호출 되어 클리어 이후 `resolve` 되는 경우에도 쿼리 클라이언트에 캐싱이 되지 않는군요 !! 

제가 잘못 이해했었나 봅니다.

클리어 해도 클리어 이후 리졸브 되는 쿼리문이 캐싱되지 않는 이유는 쿼리 클라이언트 내부 쿼리 캐시에서 저장 하고 있는 쿼리키가 하나도 존재하지 않기 때문에 해당 키를 가진 쿼리의 상태 변화는 더 이상 쿼리 클라이언트에게 영향을 끼치지 않기 때문입니다.

aken-you commented on src/app/DevTools.tsx:
```diff
@@ -62,16 +63,47 @@ export const DevTools = () => {
+      "Authorization-refresh=staleRefreshToken; path=/; max-age=3600";
+    queryClient.cancelQueries({ queryKey: ["profile", nickname] });
+    queryClient.clear();
+    queryClient.invalidateQueries({ queryKey: ["profile", nickname] });
```
clear하고 다시 무효화하는 이유가 있을까요 ?

yonghyeun replied:
제거 완료 ~!!

aken-you commented on src/entities/profile/api/dashboard.ts:
```diff
@@ -46,10 +47,23 @@ interface ProfileResponse {
+  }
+
+  const response = await fetch(PROFILE_END_POINT.PROFILE(nickname), {
+    headers: {
+      Authorization: token ?? "",
+    },
```
token이 없으면 아예 Authorization을 안보내는 건 어떨까요?

yonghyeun replied:
```tsx
export const getProfile = async ({
  nickname,
  token,
}: {
  nickname: string;
  token: AuthStore["token"];
}): Promise<ProfileResponse> => {
  const headers = new Headers();
  if (token) {
    headers.append("Authorization", token);
  }
  const response = await fetch(PROFILE_END_POINT.PROFILE(nickname), {
    headers,
  });
```

다음과 같이 수정 완료입니다

aken-you commented on src/features/auth/api/postPetInfo.ts:
```diff
@@ -95,8 +95,8 @@ export const usePostPetInfo = () => {
       queryClient.prefetchQuery({
-        queryKey: ["profile", nickname],
-        queryFn: () => getProfile(nickname as string),
+        queryKey: ["profile", { nickname, token: authorization }],
```
profile querykey에서 두 번째 key는 객체가 아닌 걸로 알고 있어요 👀 

```typescript
export const useGetProfile = ({ nickname, token }: ProfileRequest) => {
  return useQuery({
    queryKey: ["profile", nickname, token],
```

yonghyeun replied:
```tsx
    const { role, authorization: token } = data.content;
      const { lastNoneAuthRoute } = useRouteHistoryStore.getState();
      const { nickname } = useAuthStore.getState();

      setRole(role);
      setToken(token);
      navigate(lastNoneAuthRoute);

      queryClient.prefetchQuery({
        queryKey: ["profile", nickname, token],
        queryFn: nickname ? () => getProfile({ nickname, token }) : skipToken,
      });
    },
```

다음과 같이 수정 완료 했습니다

aken-you replied:
[공식 문서 skiptoken](https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#typesafe-disabling-of-queries-using-skiptoken) 링크 남길게요!
([maintainer 블로그](https://tkdodo.eu/blog/react-query-and-type-script#type-safety-with-the-enabled-option))

aken-you commented on src/mocks/handler.ts:
```diff
@@ -501,14 +527,40 @@ export const petInfoFormHandlers = [
 ];
 
+const getNewAccessTokenHandler = [
+  http.get(`${import.meta.env.VITE_API_BASE_URL}/auth`, ({ cookies }) => {
```
`/auth`는 postman에서 못봤는데, refresh token으로 access token을 재발급해주는 api url로 보면 될까요 ?

yonghyeun replied:
![image](https://github.com/user-attachments/assets/89a9ade1-ac99-416a-99c4-b157d04d8686)

넵 노션에 정의 되어 있습니다

aken-you replied:
`END_POINT.REFRESH_ACCESS_TOKEN` 상수를 사용하는 건 어떨까요?
만약 경로가 수정되면 이 상수만 수정하면 될 것 같아요 👀

yonghyeun replied:
수정 완료요

aken-you reviewed (commented):
이번에도 수고하셨습니다!

yonghyeun committed:
feat[#248] : useGetProfile skip token을 이용해 쿼리 펑션의 타입 좁히기

- https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#typesafe-disabling-of-queries-using-skiptoken

yonghyeun committed:
feat[#248] : postPetInfo onSuccess 에서 prefetch 의 쿼리 키 , 스킵 토큰 추가

yonghyeun committed:
refactor[#248] : token 을 조건부적으로 헤더에 넣도록 리팩토링

yonghyeun committed:
fix[#248] : 불필요한 invalidateQueries 제거

yonghyeun committed:
refactor[#248] : getProfile, 에러 핸들러 추가에 따라 DevTools 에서 queryClient 조작 로직 제거

yonghyeun committed:
refactor[#248] : 불필요한 주석 제거

yonghyeun committed:
fix[#248] : 에러 핸들러에서 credential same-origin 으로 변경

yonghyeun committed:
fix[#248] : msw 핸들러 변화에 따른 petInfoForm 테스트 코드 변경

yonghyeun committed:
refactor[#248] : 엔드포인트 변수명 APP_END_POINT 로 변경

yonghyeun committed:
refactor[#248] : getNewAccessToken 핸들러 상수 처리
